File: bot/models/order.py
"""
Order and OrderItem models
"""

from datetime import datetime, UTC
import enum
from .database import add_record, update_record, find_record, get_all_records


class OrderStatus(enum.Enum):
    PENDING = "pending"
    PAID = "paid"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class Order:
    def __init__(self, id=None, user_id=None, total_amount=0.0, status='pending',
                 created_at=None, completed_at=None):
        self.id = int(id) if id else None
        self.user_id = int(user_id) if user_id else None
        self.total_amount = float(total_amount) if total_amount else 0.0
        self.status = str(status).lower() if status else 'pending'
        self.created_at = str(created_at) if created_at else datetime.now(UTC).isoformat()
        self.completed_at = str(completed_at) if completed_at else None
        self.items = []
    
    def save(self):
        """Save order to Google Sheets"""
        data = {
            'user_id': self.user_id,
            'total_amount': self.total_amount,
            'status': self.status,
            'created_at': self.created_at,
            'completed_at': self.completed_at or ''
        }
        
        if self.id:
            data['id'] = self.id
            update_record('Orders', self.id, data)
        else:
            result = add_record('Orders', data)
            self.id = result['id']
        return self
    
    def get_items(self):
        """Get order items"""
        records = get_all_records('OrderItems')
        self.items = []
        for r in records:
            if str(r.get('order_id', '')) == str(self.id):
                self.items.append(OrderItem(
                    id=r.get('id'),
                    order_id=r.get('order_id'),
                    product_id=r.get('product_id'),
                    product_name=r.get('product_name'),
                    quantity=r.get('quantity', 0),
                    price=r.get('price', 0.0),
                    subtotal=r.get('subtotal', 0.0)
                ))
        return self.items
    
    @staticmethod
    def get_by_id(order_id):
        """Get order by ID"""
        _, record = find_record('Orders', 'id', order_id)
        if record:
            return Order(
                id=record.get('id'),
                user_id=record.get('user_id'),
                total_amount=record.get('total_amount', 0.0),
                status=record.get('status', 'pending'),
                created_at=record.get('created_at'),
                completed_at=record.get('completed_at')
            )
        return None
    
    @staticmethod
    def get_all_completed():
        """Get all completed orders"""
        records = get_all_records('Orders')
        orders = []
        for r in records:
            if r.get('status') == 'completed':
                orders.append(Order(
                    id=r.get('id'),
                    user_id=r.get('user_id'),
                    total_amount=r.get('total_amount', 0.0),
                    status=r.get('status', 'pending'),
                    created_at=r.get('created_at'),
                    completed_at=r.get('completed_at')
                ))
        return orders

    @staticmethod
    def get_all_pending():
        """Get all open orders (pending or paid)"""
        records = get_all_records('Orders')
        pending_orders = []
        for r in records:
            status = str(r.get('status', '')).lower()
            if status in (OrderStatus.PENDING.value, OrderStatus.PAID.value):
                pending_orders.append(Order(
                    id=r.get('id'),
                    user_id=r.get('user_id'),
                    total_amount=r.get('total_amount', 0.0),
                    status=r.get('status', 'pending'),
                    created_at=r.get('created_at'),
                    completed_at=r.get('completed_at')
                ))
        return pending_orders
    
    def __repr__(self):
        return f"<Order(id={self.id}, user_id={self.user_id}, total={self.total_amount})>"


class OrderItem:
    def __init__(self, id=None, order_id=None, product_id=None, product_name=None,
                 quantity=0, price=0.0, subtotal=0.0):
        self.id = int(id) if id else None
        self.order_id = int(order_id) if order_id else None
        self.product_id = int(product_id) if product_id else None
        self.product_name = str(product_name) if product_name else ''
        self.quantity = int(float(quantity)) if quantity else 0
        self.price = float(price) if price else 0.0
        self.subtotal = float(subtotal) if subtotal else 0.0
    
    def save(self):
        """Save order item to Google Sheets"""
        data = {
            'order_id': self.order_id,
            'product_id': self.product_id,
            'product_name': self.product_name,
            'quantity': self.quantity,
            'price': self.price,
            'subtotal': self.subtotal
        }
        
        if self.id:
            data['id'] = self.id
            update_record('OrderItems', self.id, data)
        else:
            result = add_record('OrderItems', data)
            self.id = result['id']
        return self
    
    def __repr__(self):
        return f"<OrderItem(id={self.id}, product='{self.product_name}', qty={self.quantity})>"

File: src/models/order.ts
/**
 * Order and OrderItem models
 */

import { addRecord, updateRecord, findRecord, getAllRecords } from './database';

export enum OrderStatus {
  PENDING = 'pending',
  PAID = 'paid',
  COMPLETED = 'completed',
  CANCELLED = 'cancelled'
}

export interface IOrder {
  id?: number;
  user_id: number;
  total_amount?: number;
  status?: string;
  created_at?: string;
  completed_at?: string | null;
}

export interface IOrderItem {
  id?: number;
  order_id: number;
  product_id: number;
  product_name: string;
  quantity: number;
  price: number;
  subtotal: number;
}

export class OrderItem {
  id?: number;
  order_id: number;
  product_id: number;
  product_name: string;
  quantity: number;
  price: number;
  subtotal: number;

  constructor(data: IOrderItem) {
    this.id = data.id ? parseInt(String(data.id)) : undefined;
    this.order_id = parseInt(String(data.order_id));
    this.product_id = parseInt(String(data.product_id));
    this.product_name = String(data.product_name);
    this.quantity = parseInt(String(data.quantity || 0));
    this.price = parseFloat(String(data.price || 0));
    this.subtotal = parseFloat(String(data.subtotal || 0));
  }

  /**
   * Save order item to Google Sheets
   */
  async save(): Promise<OrderItem> {
    const data = {
      order_id: this.order_id,
      product_id: this.product_id,
      product_name: this.product_name,
      quantity: this.quantity,
      price: this.price,
      subtotal: this.subtotal
    };

    if (this.id) {
      await updateRecord('OrderItems', this.id, { id: this.id, ...data });
    } else {
      const result = await addRecord('OrderItems', data);
      this.id = result.id;
    }

    return this;
  }
}

export class Order {
  id?: number;
  user_id: number;
  total_amount: number;
  status: string;
  created_at: string;
  completed_at: string | null;
  items: OrderItem[];

  constructor(data: IOrder) {
    this.id = data.id ? parseInt(String(data.id)) : undefined;
    this.user_id = parseInt(String(data.user_id));
    this.total_amount = parseFloat(String(data.total_amount || 0));
    this.status = String(data.status || OrderStatus.PENDING).toLowerCase();
    this.created_at = data.created_at || new Date().toISOString();
    this.completed_at = data.completed_at || null;
    this.items = [];
  }

  /**
   * Save order to Google Sheets
   */
  async save(): Promise<Order> {
    const data = {
      user_id: this.user_id,
      total_amount: this.total_amount,
      status: this.status,
      created_at: this.created_at,
      completed_at: this.completed_at || ''
    };

    if (this.id) {
      await updateRecord('Orders', this.id, { id: this.id, ...data });
    } else {
      const result = await addRecord('Orders', data);
      this.id = result.id;
    }

    return this;
  }

  /**
   * Get order items
   */
  async getItems(): Promise<OrderItem[]> {
    const records = await getAllRecords('OrderItems');
    this.items = [];

    for (const record of records) {
      if (String(record.order_id) === String(this.id)) {
        this.items.push(new OrderItem(record as IOrderItem));
      }
    }

    return this.items;
  }

  /**
   * Get order by ID
   */
  static async getById(orderId: number): Promise<Order | null> {
    const [, record] = await findRecord('Orders', 'id', orderId);
    return record ? new Order(record as IOrder) : null;
  }

  /**
   * Get all completed orders
   */
  static async getAllCompleted(): Promise<Order[]> {
    const records = await getAllRecords('Orders');
    return records
      .filter(r => r.status === OrderStatus.COMPLETED)
      .map(r => new Order(r as IOrder));
  }

  /**
   * Get all pending orders
   */
  static async getAllPending(): Promise<Order[]> {
    const records = await getAllRecords('Orders');
    return records
      .filter(r => r.status === OrderStatus.PENDING || r.status === OrderStatus.PAID)
      .map(r => new Order(r as IOrder));
  }
}
